docs/src/dev/langref.md

```diff
@@ -1,4 +1,286 @@
+
+
+### Cell identifiers {#cell}
+
+```
+%<index>
+%<index>:<width>
+%<index>+<offset>
+%<index>+<offset>:<width>
+%<index>:_
+```
+
+For example:
+
+```
+%0
+%1:4
+%1+0
+%1+2:2
+%2:0
+%3:_
+```
+
+Cell identifiers start with a `%` sigil followed by a non-empty sequence of `0`..`9` digits, optionally followed by a `+` character and a non-empty sequence of `0`..`9` digits (denoting an offset into the wide output of a cell), optionally followed by a `:` character and a non-empty sequence of `0`..`9` digits (denoting the width of a cell declaration or reference). No whitespace is allowed between any of the characters.
+
+A *placeholder* cell identifier starts with a `%` sigil followed by a non-empty sequence of `0`..`9` digits, a `:` character, and a `_` character. No whitespace is allowed between any of the characters. The placeholder cell identifier is only used when declaring cells with multiple outputs.
+
+<div class="warning">
+
+While the in-memory IR has the property that a cell with a wide output occupies a contiguous range of cell indices (i.e. that `%0+1` and `%1` always refer to the same net), the same is not true in the text IR. Instead, each unique `<index>` can refer to a distinct cell with an output of any width.
+
+To avoid confusion, the IR printer ensures that in references, `<index>` is equal to the lowest index occupied by a cell with a wide output (i.e. it will never format `%0+1` as `%1`).
+
+</div>
+
+
+### Concatenations {#concat}
+
+```
+[ <part>... ]
+```
+
+For example:
+
+```
+[]
+[&"gpio":1 &"gpio":0]
+[&"pin" &_]
+[1001]
+[10 01]
+[%0 %1:4 000]
+[ XXXX ]
+```
+
+Concatenations are delimited by `[` and `]` characters on either end and contain zero or more parts delimited by [whitespace](#whitespace). The first part of the concatenation in the source code order corresponds to the most significant positions of the resulting reference. The two types of concatenations are *I/O concatenations* and *value concatenations*.
+
+Each part of an I/O concatenation must be an [I/O identifiers](#io), which could be named or floating.
+
+Each part of a value concatenation must be a [constant](#constant) or a [cell identifier](#cell).
+
+
+### Other special characters {#special}
+
+Special characters `[`, `]`, `(`, `)`, `{`, `}`, `=`, and `,` are used as delimiters within [declarations](#decls).
+
+These characters may be optionally surrounded by the [whitespace](#whitespace) characters `U+0020 SPACE` and `U+0009 CHARACTER TABULATION` without altering their meaning. In addition, the [whitespace](#whitespace) character `U+000A LINE FEED` may appear between the matching `[`, `]`, `(`, `)`, `{`, or `}` delimiter pairs, in which case it does not indicate the end of the syntactic construct.
+
+
+## Header {#header}
+
+The header must be located at the beginning of the file, excluding any prior [whitespace](#whitespace) or [comments](#comment). It describes how the remainder of the file must be interpreted.
```

Comment:
I don't have better wording on hand but "excluding" is ambiguous as to whether it describes the sane option or the sucky option.